File: A3/A3-025/A3-025.cpp
#include<iostream>
#include<vector>
using namespace std;

bool hasAllHole = true;
int N, W, L;
vector<vector<bool>> woods;

int main()
{
    // This makes std::cin and std::cout faster
    ios_base::sync_with_stdio(false); cin.tie(0);
    // Input
    cin >> N >> W >> L;
    woods.assign(N, vector<bool>(W, false));

    // Input position of the hole
    for(int i = 0; i < N; i++)
    {
        int K;
        cin >> K;
        while(K--)
        {
            int hole;
            cin >> hole;
            hole--;

            // Mark [hole - L, hole + L] as a hole
            for(int dl = -L; dl <= L; dl++)
            {
                // Check out of bounds
                if(hole + dl < 0 || hole + dl >= W) { continue; }
                // Mark as a hole
                woods[i][hole + dl] = true;
            }
        }
    }

    // Check hole on all woods
    for(int x = 0; x < W; x++)
    {
        hasAllHole = true;
        for(int i = 0; i < N; i++)
        {
            if(!woods[i][x])
            {
                hasAllHole = false;
                break;
            }
        }
        if(hasAllHole) { break; }
    }

    // Output
    if(hasAllHole) { cout << "1\n"; }
    else { cout << "0\n"; }
    return 0;
}

File: A3/A3-025/A3-025.py
# Input
N, W, L = map(int, input().split())

# Input position of the hole
woods = [[False] * W for _ in range(N)]
for i in range(N):
    holes = list(map(int, input().split()))
    holes = holes[1:]
    # Mark [hole - L, hole + L] as a hole
    for hole in holes:
        for dl in range(-L, L + 1):
            if(0 <= hole + dl < W):
                woods[i][hole + dl] = True

# Check hold on all woods
hasAllHoles = True
for x in range(W):
    hasAllHoles = True
    for i in range(N):
        if(not woods[i][x]):
            hasAllHoles = False
            break
    if(hasAllHoles):
        break

# Output
if(hasAllHoles):
    print(1)
else:
    print(0)